ceph/ceph
# mgr/dashboard: Add column filtering feature to cd-table

Add column filtering feature to the cd-table component.
- Create filters from columns.
- Create filter options from row data in columns.
- Allow specifying custom functions to do filtering.
- Allow specifying hard-coded options for filters.
- Allow specifying extra filters other than visible columns on table.

Previously, this feature was implemented in inventory component. Refactor
it to move the logic into cd-table, so all tables can use the feature if
needed.

Also, the previous implementation occupied too much space in the table header if there are many filters.

Signed-off-by: Kiefer Chang <kiefer.chang@suse.com>
## Usage
- We can enable the filtering feature for a column by enable `filterable` attribute:
```typescript
    this.columns = [
      {
        prop: 'pool_name',
        name: this.i18n('Name'),
        flexGrow: 4,
        cellTransformation: CellTemplate.executing
      },
      {
        prop: 'type',
        name: this.i18n('Type'),
        flexGrow: 2,
        filterable: true // turn on the filtering feature for this column
      },
```
- We can also add some extra filters and custom filter functions: like https://github.com/bk201/ceph/commit/d7f4d0e69063e46017632fbbe2df3b362d25d8b7

## How to test
- Check **Cluster-->Inventory** page by loading some fake data
```
cd /ceph/build
bin/ceph mgr module enable test_orchestrator
bin/ceph orchestrator set backend test_orchestrator
wget https://raw.githubusercontent.com/bk201/ceph-resources/master/issues/40335/dummy.json -O /tmp/dummy.json
bin/ceph test_orchestrator load_data -i /tmp/dummy.json
```
- Check **Cluster-->Configuration** page
## Screenshots
The inventory page is adapted to this change:
![inventory](https://user-images.githubusercontent.com/1691518/72058125-8c044100-330a-11ea-87b4-937ae9a20dd2.gif)

Also the configuration page:
![configuration](https://user-images.githubusercontent.com/1691518/72058378-f4ebb900-330a-11ea-83bd-1e8cfbd30cd5.gif)



## Checklist

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>
